buildpacks/nodejs-npm-install/src/main.rs

```diff
@@ -0,0 +1,212 @@
+mod errors;
+mod layers;
+mod npm;
+
+use crate::errors::NpmInstallBuildpackError;
+use crate::layers::npm_cache::NpmCacheLayer;
+use commons::fun_run::CommandWithName;
+use commons::output::build_log::{BuildLog, Logger, SectionLogger};
+use commons::output::fmt;
+use commons::output::section_log::{log_step, log_step_stream};
+use commons::output::warn_later::WarnGuard;
+use heroku_nodejs_utils::application;
+use heroku_nodejs_utils::package_json::PackageJson;
+use heroku_nodejs_utils::package_manager::PackageManager;
+use heroku_nodejs_utils::vrs::Version;
+use libcnb::build::{BuildContext, BuildResult, BuildResultBuilder};
+use libcnb::data::build_plan::BuildPlanBuilder;
+use libcnb::data::launch::{LaunchBuilder, ProcessBuilder};
+use libcnb::data::{layer_name, process_type};
+use libcnb::detect::{DetectContext, DetectResult, DetectResultBuilder};
+use libcnb::generic::{GenericMetadata, GenericPlatform};
+use libcnb::{buildpack_main, Buildpack, Env};
+use std::io::{stdout, Stdout};
+use std::path::Path;
+use std::process::Command;
+
+pub(crate) struct NpmInstallBuildpack;
+
+impl Buildpack for NpmInstallBuildpack {
+    type Platform = GenericPlatform;
+    type Metadata = GenericMetadata;
+    type Error = NpmInstallBuildpackError;
+
+    fn detect(&self, context: DetectContext<Self>) -> libcnb::Result<DetectResult, Self::Error> {
+        context
+            .app_dir
+            .join(PackageManager::Npm.lockfile())
+            .exists()
+            .then(|| {
+                DetectResultBuilder::pass()
+                    .build_plan(
+                        BuildPlanBuilder::new()
+                            .provides("node_modules")
+                            .requires("npm")
+                            .requires("node_modules")
+                            .requires("node")
+                            .build(),
+                    )
+                    .build()
+            })
+            .unwrap_or_else(|| DetectResultBuilder::fail().build())
+    }
+
+    fn build(&self, context: BuildContext<Self>) -> libcnb::Result<BuildResult, Self::Error> {
+        let logger = BuildLog::new(stdout()).buildpack_name("Heroku npm Engine Buildpack");
+        let warn_later = WarnGuard::new(stdout());
+        let env = Env::from_current();
+        let app_dir = &context.app_dir;
+        let package_json = PackageJson::read(app_dir.join("package.json"))
+            .map_err(NpmInstallBuildpackError::PackageJson)?;
+
+        run_application_checks(app_dir, &warn_later)?;
+
+        let section = logger.section("Installing node modules");
+        log_npm_version(&env, section.as_ref())?;
+        configure_npm_cache_layer(&context, &env, section.as_ref())?;
+        run_npm_install(&env, section.as_ref())?;
+        let logger = section.end_section();
+
+        let section = logger.section("Running scripts");
+        run_build_scripts(&package_json, &env, section.as_ref())?;
+        let logger = section.end_section();
```

Comment:
Why redeclare this variable when you it could have been made mutable on the first declaration?

Reply:
No specific reason, I just prefer to avoid `mut` if it's not necessary and a `let` reassignment works just as well here.